### PR TITLE
Float8 autoquant weight only

### PR DESCRIPTION
**Pull request by jainapurva**
Added weight-only float8 quantization technique to autoquant. It runs the default kernel where dequantize is performed on float8 weight tensor followed by linear.

Updated the safe_int_mm to work on H100 for float8 dtypes.

**Test Results**

Llama 3.1 8b on H100

**Int8 Autoquant - weight only**

Float8 autoquant, with float16 and compile

- 	Time for inference 5: 0.99 sec total
- 	Average tokens/sec: 201.71
- 	Average Bandwidth: 1519.97 GB/s
- 	Peak Memory Usage: 12.04 GB
- 	Model Size: 7.54 GB

Autoquant for float16 model and compile - int8wo

- 	Time for inference 5: 0.98 sec total
- 	Average tokens/sec: 203.48
- 	Average Bandwidth: 1535.56 GB/s
- 	Peak Memory Usage: 11.10 GB
- 	Model Size: 7.55 GB

### CODE REVIEW

File: scripts/hf_eval.py [hunk 1]
```diff
@@ -89,7 +89,7 @@ def all_linear(mod, name):
     with torch.no_grad():
         result = evaluate(
             HFLM(
-                pretrained=model.to(device),
+                pretrained=model,
```
Comment (jerryzh168) [inline scripts/hf_eval.py:92]: is this change needed

File: scripts/prepare.sh
Comment (jerryzh168) [inline scripts/prepare.sh:1]: empty file?

File: scripts/hf_eval.py (hunk 1 above)
Comment (jainapurva) [reply scripts/hf_eval.py]: This changeis removed from current PR, will be done in separate PR.

File: scripts/prepare.sh
Comment (jainapurva) [reply scripts/prepare.sh:1]: Removed

**Review by HDCharles — CHANGES_REQUESTED** [review]
some additional things you should add:


1) this option should probably detect the type of gpu you're using and only add float8 if its relevant
2) make sure if you run it on H100 for Llama that when the op in question is chosen, it actually speeds things up
3) add to CI like you mention, you can see the tests in the PR: 
https://github.com/pytorch/ao/pull/804/files

**Comment by jainapurva** [comment]
> some additional things you should add:
> 
> 1. this option should probably detect the type of gpu you're using and only add float8 if its relevant
> 2. make sure if you run it on H100 for Llama that when the op in question is chosen, it actually speeds things up
> 3. add to CI like you mention, you can see the tests in the PR:
>    https://github.com/pytorch/ao/pull/804/files

As discussed offline. I've added the autoquant in a separate list and architecture check. New kernels need to be added in future to improve the time. The current time is equivalent with int8.

**Review by HDCharles — APPROVED** [review]
looks good

nit: I would give a full list of updates in your PR description if possible
i.e. you also add H100 support for safe_int_mm

nit: it would probably be good to have a unit test for safe_int_mm on H100 even if it has to be run manually

File: torchao/kernel/intmm.py [hunk 2]
```diff
@@ -69,7 +69,10 @@ def safe_int_mm(input: torch.Tensor, mat2: torch.Tensor) -> torch.Tensor:
             input = (
                 input.contiguous()
             )  # (it seems the transpose makes cublas check the above j constraint on i)
-        return out_dtype(torch.ops.aten.mm.default, torch.int32, input, mat2)
+        try:
```
Comment (jerryzh168) [inline torchao/kernel/intmm.py:72]: maybe adding a comment to this would be helpful, how these two branches are handled?

Comment (jainapurva) [reply torchao/kernel/intmm.py:72]: The except is executed if it's a float8 dtype on H100, as there's no implementation for addmm_cuda for float8 dtypes
